### DISCUSSION TITLE
Publish with "Trim unused code"

### DISCUSSION THREAD

**Discussion by KoalaBear84** [Tyrrrz/CliFx#141] [Q&A] [answered]
I'm trying to publish a simple console app with CliFx and a single `DefaultCommand`.

When trimmed it's about 12-13MB, but not working. It returns the error, only when specifically using `.AddCommand<DefaultCommand>()`:
```
Failed to create an instance of type `MyConsoleApp.Commands.DefaultCommand`, could not access the constructor.
Default type activator is only capable of instantiating a type if it has a public parameterless constructor.
To fix this, either add a parameterless constructor to the type or configure a custom activator for the application.
```

When only using the following, it doesn't see the DefaultCommand at all, probably because it discards it before adding it:
```csharp
    .SetExecutableName("app")
    .AddCommandsFromThisAssembly()
```

Already tried the following:
```csharp
[Command]
[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
public class DefaultCommand : ICommand
{
    public DefaultCommand()
    {
        
    }

    [CommandOption("input-path", 'p', Description = "Input Path")]
    public string InputPath { get; set; } = string.Empty;

    ...
}
```

Source:
https://devblogs.microsoft.com/dotnet/customizing-trimming-in-net-core-5/
https://github.com/dotnet/sdk/issues/36835

Anyone has more info about this?

**Comment by Tyrrrz** [comment, marked as answer]
CliFx is not trim-compatible in its current state and it would take a lot of effort to make it compatible (essentially rewriting it with source-generators), due to the heavy use of type-walking and dynamic type conversions.

In order to get your CLI app working with CliFx with assembly trimming, you have to manually preserve your commands, converters, and other dynamically referenced members (such as `Parse(...)` methods on applicable types, when used for conversion).

See here for a working example: https://github.com/Tyrrrz/DiscordChatExporter/blob/c8ea365c04498bea973c38e2ed028e8c31a05bb0/DiscordChatExporter.Cli/Program.cs#L13-L40. Feel free to narrow the `DynamicallyAccessedMemberTypes` to trim more code, generally only the `PublicConstructors`, `PublicMethods`, and `Interfaces` are required.

**Reply by KoalaBear84** [reply]
Many thanks, I'll try it ASAP and report back 👍

**Reply by KoalaBear84** [reply]
Many thanks again!

For me it was as simple as adding this single line because I only have one command:
```csharp
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(DefaultCommand))]
    private static async Task<int> Main(string[] args)
```
